kratos/elements/distance_calculation_flux_based_element.cpp

```diff
@@ -0,0 +1,502 @@
+        const ProcessInfo &rCurrentProcessInfo)
+{
+    BoundedVector<double,TNumNodes> gauss_weights;
+    BoundedMatrix<double,TNumNodes,TNumNodes> N_container;
+    array_1d<BoundedMatrix<double, TNumNodes, TDim>, TNumNodes> DN_DX_container;
+
+    CalculateGaussPointsData( GetGeometry(), gauss_weights, N_container, DN_DX_container);
+
+    array_1d<double, NumNodes > nodal_values;
+    array_1d< array_1d<double, TDim >, NumNodes> v; //convection velocity
+    array_1d< array_1d<double, TDim >, NumNodes> v_unit; //to decide if convection must be turned off
+    array_1d<double, TDim > avg_v_unit = ZeroVector(TDim); //to decide if convection must be turned off
+    bool has_fixed_node = false;
+
+    for(unsigned int i=0; i<NumNodes; i++){
+        nodal_values[i] = GetGeometry()[i].FastGetSolutionStepValue(DISTANCE);
+        const auto& rVel_i = GetGeometry()[i].GetValue(POTENTIAL_GRADIENT);
+        for(unsigned int j=0; j<TDim; j++){
+            v[i][j] = rVel_i[j];
+        }
+        v_unit[i] =  v[i]/norm_2(v[i]); // |v_unit| = 1
+        avg_v_unit += v_unit[i];
+        has_fixed_node = has_fixed_node || GetGeometry()[i].IsFixed(DISTANCE);
+    }
+    avg_v_unit/=static_cast<double>(NumNodes);
+
+    BoundedMatrix<double, TNumNodes, TDim> avg_DN_DX = ZeroMatrix(TNumNodes, TDim);
+    for(unsigned int i=0; i<TNumNodes; i++){
+        avg_DN_DX += DN_DX_container[i] /static_cast<double>(TNumNodes);
+    }
+    
+    //adjusting RHS to get |grad|=1
+    if(rCurrentProcessInfo[NL_ITERATION_NUMBER] > 1) {
+        const array_1d<double, TDim> avg_grad = prod(trans(avg_DN_DX), nodal_values);
+        const double grad_modulus = norm_2(avg_grad);
+
+        const double new_correction_coeff = 1.0/grad_modulus;
+        const double relaxation_factor = 0.5;
+        mCorrectionCoefficient = mCorrectionCoefficient *(1.0-relaxation_factor) + new_correction_coeff*relaxation_factor;
+    }
+    else {
+        mCorrectionCoefficient = 1.0;
+    }
+
+    //computing element size(for Tau)
+    const double h = ComputeH(avg_DN_DX);
+
+    //DECIDE IF CONVECTION IS ACTIVE/INACTIVE
+   //we decide by checking if flow converges to /diverges from this element
+   //if all velocity vectors are oriented in the same direction, then convection is valid
+   //but if the element acts as a "sink" or "source", convection is turned off to avoid instabiliities.
+    bool add_convection = true;
+    double average_unit_vel_misaligment = 0.0;
+    for(unsigned int i=0; i<NumNodes; i++) {
+        average_unit_vel_misaligment += norm_2( v_unit[i]  - avg_v_unit );
+    }
+    average_unit_vel_misaligment/=static_cast<double>(NumNodes);
+    if(average_unit_vel_misaligment>0.33) add_convection=false;
+
+    //checking if the flow comes from face that has neighbour element (only if it is not inlet)
+    if( !has_fixed_node && this->Has(NEIGHBOUR_ELEMENTS) && this->GetValue(NEIGHBOUR_ELEMENTS).size()==GetGeometry().FacesNumber())
```

Comment:
```suggestion
    if( !has_fixed_node && this->GetValue(NEIGHBOUR_ELEMENTS).size()==GetGeometry().FacesNumber())
```
As far as I understand this requires the `NEIGHBOUR_ELEMENTS` to be computed. Then I think we can assume that these are already there by either calling the neighbors calculation or checking it in the `FluxBasedRedistanceProcess`.

Reply:
ok, moved it to the check